mono/mini/debugger-agent.c

```diff
@@ -6009,12 +6011,43 @@ clear_types_for_assembly (MonoAssembly *assembly)
 	mono_loader_unlock ();
 }
 
+//Count threads removing the thread of finalizer GC if it is not executing a managed code
+static void
+count_thread_checking_finalizer_gc (gpointer key, gpointer value, gpointer remove_finalizer)
+{
+	gboolean *ret = (gboolean *)remove_finalizer;
+	MonoThread *thread = (MonoThread *)value;
+	if (mono_gc_is_finalizer_internal_thread(thread->internal_thread)) {
+		DebuggerTlsData *tls = (DebuggerTlsData *)mono_g_hash_table_lookup (thread_to_tls, thread->internal_thread);
+		if (tls->frame_count == 0) {
+			*ret = TRUE;
+			return;
+		}
+		if (tls->frame_count) {
+			StackFrame* frame = tls->frames [0];
```

Comment:
I think the behavior we want is:
* if the GC Finalizer thread is idle, don't show it in the thread list.
* if the GC Finalizer is running some finalizer code, show it.

I'm not sure that the code here does that.  I think `frames [0]` is the currently executing function, right?  That means if I have a finalizer that calls some native code, we won't show the finalizer in the debugger.

I wonder if it would be better if the finalizer toggled some flag.  We raise two profiler events around the place where we actually run the managed finalizers:
https://github.com/mono/mono/blob/96b1fd837d6caf1be1b657c0726af3b8caae7060/mono/metadata/gc.c#L925-L932

I think the debugger could add callbacks for those two events to toggle some private bit so that it always knows if any managed finalizers are currently running in the finalizer thread.

Reply:
Thanks for the suggestion! Done!